### PR TITLE
Track popup to cursor [ready]

### PR DESCRIPTION
**Pull request by peterqliu**
ticket: https://github.com/mapbox/mapbox-gl-js/issues/7777

Adds a method `.trackPointer()` to the Popup API that repositions the popup to the mouse cursor on every mousemove event, and hides/shows the popup when cursor leaves/enters the map. Mutually overwritable with `.setLngLat`.

### Design

Opted not to extend setLngLat, as it doesn't quite fit those words. Also opted not to extend the popup `options` object, for cases when users need to change behavior after adding to map.

 ## Launch Checklist

 - [x] briefly describe the changes in this PR
 - [x] write tests for all new functionality
 - [x] document any changes to public APIs
 - [x] manually test the debug page
 - [x] tagged `@mapbox/studio` and/or `@mapbox/maps-design` if this PR includes style spec changes

### CODE REVIEW

**Comment by ryanhamley** [comment]
> Touchscreen behavior: show popup only on touchmove? Not show at all? Console warning when invoking .trackCursor() on touch devices?

I'm trying to imagine a use case for this on mobile. It seems like it would be a very awkward UX as compared to simply touching on each new point to see a new popup. On the other hand, having a desktop only API method seems odd. Do we have other methods that don't work on mobile? If we do enable this on `touchmove`, maybe it would be better to call the method a more generic term like `trackPointer`.

> Close button: won't be clickable when popup shadows mouse. Should we hide? Does that affect any corner cases?

This is a fun consequence. Maybe there needs to be a separate action to close the popup in this case. Would clicking the map to toggle the popup be too weird?

**Comment by peterqliu** [comment]
#### Mobile handling

In lieu of hover, I thought about adding cursor-tracked popups to a tap on the map. The interaction pattern for mobile is unfortunately quite complex: `click` is the event that typically corresponds to this, but is nondeterministic. On Webkit browsers where the user has also implemented a `hover` listener, `click` is invoked only on the *second* tap, and not a double-tap at that. Because of this unpredictability, I've removed touchscreen support entirely for the time being.

#### Other changes

- renamed `trackCursor` to `trackPointer`, to reflect the mouse-specific compatibility
- move the show/hide logic to CSS, greatly simplifying the event handler logic
- opted to keep the closeButton by default, as is the current Popup default. The change is easy enough on the user's side

**Comment by peterqliu** [comment]
@KravMaguy that is correct, you will lose that functionality if you use a cursor-tracked popup on mobile. In these contexts, most apps display the information in a panel separate from the map: it's a much better practice given the limited screen space, and preserves visibility/interactivity of the map itself.

some examples ⬇️ 

<details>

img_3115
img_3117
img_3116

</details>

**Comment by peterqliu** [comment]
@KravMaguy you can still `unproject`/`setLngLat` the popup on each touch event, to move the popup there

File: src/ui/popup.js [hunk 1]
```diff
@@ -95,13 +96,18 @@ export default class Popup extends Evented {
...
             this._map.on('click', this._onClickClose);
         }
+
         this._map.on('remove', this.remove);
         this._update();
 
+        if (this._trackPointer) {
+            this._map.on('mousemove', (e) => { this._update(e.point); });
+            this._container.classList.add('mapboxgl-popup-track-pointer');
+        } else this._map.on('move', this._update);
```
Comment (ryanhamley) [inline src/ui/popup.js]: I'd wrap the else block in brackets

**Comment by ryanhamley** [comment]
> Agree that this is weird, but the specific solution seems like it would vary by user/use case/preference. I'm inclined to let people modify CSS as appropriate.

👍 I do think it makes sense to change the cursor from a grabber to a pointer when in `trackPointer` mode. The idea of this feature is to point at various locations on the map and get information about them, so the pointer feels semantically valid as well as it visually covering less of the popup.

**Comment by peterqliu** [comment]
- opted to hide popup when dragging the map -- tracking popup on drag made it harder to see the map when repositioning it takes priority

- cursor is now `pointer` when tracking popup

**Comment by peterqliu** [comment]
>If you move the mouse quickly enough, you can end up inside the popup. I'm not entirely sure what the answer here could be, if anything.

Also removed interactivity with pointer-tracked popups, to avoid getting trapped in it 👍

**Comment by Plantain** [comment]
Is this likely to make it into a release in the near future? We'd like to use this feature rather than the workarounds in #7777

**Comment by peterqliu** [comment]
@Plantain this branch is ready to the extent that I can tell -- looping in some reviewers to see if we can get this across the line.

File: src/ui/popup.js [hunk 2]
```diff
@@ -283,18 +322,18 @@ export default class Popup extends Evented {
...
             this._closeButton.innerHTML = '&#215;';
             this._closeButton.addEventListener('click', this._onClickClose);
         }
+
     }
 
-    _update() {
-        if (!this._map || !this._lngLat || !this._content) { return; }
+    _update(cursor: PointLike) {
+        if (!this._map || !this._lngLat && !this._trackPointer || !this._content) { return; }
```
Comment (ryanhamley) [inline src/ui/popup.js]: for clarity's sake, you may want to explicitly group the middle term here: `!this._map || (!this._lngLat && !this._trackPointer) || !this._content`

**Comment by ryanhamley** [comment]
This is too late to be in the next release of GL JS which comes out this week, but it should be in May's release

**Comment by peterqliu** [comment]
clarified a bit of the code per https://github.com/mapbox/mapbox-gl-js/pull/7786#issuecomment-479154849 and squashed as much as I can (not sure if I can squash past a previous pull from master?)

**Comment by ryanhamley** [comment]
You don't have to squash ahead of time @peterqliu. Once all the checks have passed, just pressing the `Squash and merge` button will take care of everything.

**Pull request merged**